File: app/src/main/java/io/plan8/backoffice/fragment/MoreFragment.java
package io.plan8.backoffice.fragment;

import android.database.Cursor;
import android.databinding.DataBindingUtil;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import io.plan8.backoffice.ApplicationManager;
import io.plan8.backoffice.BR;
import io.plan8.backoffice.R;
import io.plan8.backoffice.SharedPreferenceManager;
import io.plan8.backoffice.adapter.RestfulAdapter;
import io.plan8.backoffice.databinding.FragmentMoreBinding;
import io.plan8.backoffice.listener.EndlessRecyclerOnScrollListener;
import io.plan8.backoffice.model.BaseModel;
import io.plan8.backoffice.model.api.Attachment;
import io.plan8.backoffice.model.api.Member;
import io.plan8.backoffice.model.api.User;
import io.plan8.backoffice.model.item.EmptySpaceItem;
import io.plan8.backoffice.model.item.LabelItem;
import io.plan8.backoffice.vm.MoreFragmentVM;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by SSozi on 2017. 11. 28..
 */

public class MoreFragment extends BaseFragment {
    private final int GET_USER = 1;
    private final int GET_MEMBERS = 2;
    private boolean isCompletedGetUser = false;
    private boolean isCompletedGetMembers = false;
    private FragmentMoreBinding binding;
    private MoreFragmentVM vm;
    private List<BaseModel> moreFragmentData;
    private EndlessRecyclerOnScrollListener endlessRecyclerOnScrollListener;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(LayoutInflater.from(getContext()), R.layout.fragment_more, container, false);
        vm = new MoreFragmentVM(this, savedInstanceState);
        binding.setVariable(BR.vm, vm);
        binding.executePendingBindings();
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        endlessRecyclerOnScrollListener = new EndlessRecyclerOnScrollListener() {
            @Override
            public void onLoadMore(int currentPage) {
                refreshMoreFragmentData();
            }
        };

        binding.moreFragmentRefresh.setColorSchemeResources(R.color.colorAccent);
        binding.moreFragmentRefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                moreFragmentData.clear();
                endlessRecyclerOnScrollListener.initPrevItemCount();
                refreshMoreFragmentData();
                binding.moreFragmentRefresh.setRefreshing(false);
            }
        });
        refreshMoreFragmentData();
    }

    @Override
    public void onDestroy() {
        if (null != binding) {
            binding.unbind();
        }
        super.onDestroy();
    }

    private void completeRefreshData(int apiFlag) {
        if (null == moreFragmentData) {
            moreFragmentData = new ArrayList<>();
        }
        if (apiFlag == GET_USER) {
            isCompletedGetUser = true;
        }
        if (apiFlag == GET_MEMBERS) {
            isCompletedGetMembers = true;
        }

        if (isCompletedGetMembers && isCompletedGetUser) {
            moreFragmentData = new ArrayList<>();

            moreFragmentData.add(new LabelItem("내 프로필"));
            if (ApplicationManager.getInstance().getUser() != null) {
                moreFragmentData.add(ApplicationManager.getInstance().getUser());
            }
            moreFragmentData.add(new LabelItem("내가 속한 팀"));

            if (null != ApplicationManager.getInstance().getMembers()) {
                List<Member> members = new ArrayList<>();
                for (Member m : ApplicationManager.getInstance().getMembers()) {
                    if (null != m) {
                        members.add(m);
                    }
                }

                if (members.size() > 0) {
                    moreFragmentData.addAll(members);
                }
            }
            moreFragmentData.add(new EmptySpaceItem(0));

            vm.setData(moreFragmentData);
        }
        setCompletedLoading(true);
    }

    public void refreshMoreFragmentData() {
        setCompletedLoading(false);
        Call<List<Member>> getUserMembersCall = RestfulAdapter.getInstance().getServiceApi().getUserMembers("Bearer " + SharedPreferenceManager.getInstance().getUserToken(getContext()), ApplicationManager.getInstance().getUser().getId());
        getUserMembersCall.enqueue(new Callback<List<Member>>() {
            @Override
            public void onResponse(Call<List<Member>> call, Response<List<Member>> response) {
                List<Member> members = response.body();
                ApplicationManager.getInstance().setMembers(members);
                completeRefreshData(GET_MEMBERS);
            }

            @Override
            public void onFailure(Call<List<Member>> call, Throwable t) {
                completeRefreshData(GET_MEMBERS);
            }
        });

        Call<User> meCall = RestfulAdapter.getInstance().getServiceApi().getMe("Bearer " + SharedPreferenceManager.getInstance().getUserToken(getContext()));
        meCall.enqueue(new Callback<User>() {
            @Override
            public void onResponse(Call<User> call, Response<User> response) {
                User user = response.body();
                ApplicationManager.getInstance().setUser(user);
                completeRefreshData(GET_USER);
            }

            @Override
            public void onFailure(Call<User> call, Throwable t) {
                completeRefreshData(GET_USER);
            }
        });
    }

    public void uploadImage(Uri data) {
        setCompletedLoading(false);
        Uri uri = data;
        String imagePath = getRealPathFromURI(uri);

        final Cursor cursor = getActivity().getContentResolver().query(Uri.parse(uri.toString()), null, null, null, null);
        assert cursor != null;
        cursor.moveToNext();
        String absolutePath = cursor.getString(cursor.getColumnIndex(MediaStore.MediaColumns.DATA));
        cursor.close();

        File files = new File(absolutePath);
        RequestBody requestFile = RequestBody.create(MediaType.parse(getActivity().getContentResolver().getType(uri)), files);
        MultipartBody.Part multipart = MultipartBody.Part.createFormData("files", files.getName(), requestFile);
        Call<List<Attachment>> uploadCall = RestfulAdapter.getInstance().getServiceApi().postUpload("Bearer " + SharedPreferenceManager.getInstance().getUserToken(getActivity()), multipart);
        uploadCall.enqueue(new Callback<List<Attachment>>() {
            @Override
            public void onResponse(Call<List<Attachment>> call, Response<List<Attachment>> response) {
                List<Attachment> attachments = response.body();
                if (attachments != null) {
                    HashMap<String, String> putMeImage = new HashMap<String, String>();
                    putMeImage.put("avatar", attachments.get(0).getUrl());
                    Call<User> putMe = RestfulAdapter.getInstance().getServiceApi().putMe("Bearer " + SharedPreferenceManager.getInstance().getUserToken(getActivity()), putMeImage);
                    putMe.enqueue(new Callback<User>() {
                        @Override
                        public void onResponse(Call<User> call, Response<User> response) {
                            if (response.body() != null) {
                                ApplicationManager.getInstance().setUser(response.body());
                                refreshMoreFragmentData();
                            }
                            setCompletedLoading(true);
                        }

                        @Override
                        public void onFailure(Call<User> call, Throwable t) {
                            Toast.makeText(getContext(), "프로필 사진 업로드에 실패하였습니다. 잠시 후 다시 시도해주세요.", Toast.LENGTH_SHORT).show();
                            setCompletedLoading(true);
                        }
                    });
                }
            }

            @Override
            public void onFailure(Call<List<Attachment>> call, Throwable t) {
                Toast.makeText(getContext(), "프로필 사진 업로드에 실패하였습니다. 잠시 후 다시 시도해주세요.", Toast.LENGTH_SHORT).show();
                setCompletedLoading(true);
            }
        });
    }

    private String getRealPathFromURI(Uri imageUri) {
        String[] path = {MediaStore.Images.Media.DATA};
        Cursor cursor = getActivity().managedQuery(imageUri, path, null, null, null);
        int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
        cursor.moveToFirst();
        return cursor.getString(column_index);
    }

    public void setCompletedLoading(boolean completedLoading) {
        vm.setCompletedLoading(completedLoading);
    }
}

File: app/src/main/java/io/plan8/backoffice/fragment/MoreFragment.kt
package io.plan8.backoffice.fragment

import android.annotation.SuppressLint
import android.app.Activity
import android.databinding.DataBindingUtil
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import io.plan8.backoffice.BR
import io.plan8.backoffice.Constants
import io.plan8.backoffice.R
import io.plan8.backoffice.databinding.FragmentMoreBinding
import io.plan8.backoffice.model.item.*
import io.plan8.backoffice.vm.MoreFragmentVM
import android.provider.MediaStore
import android.util.Log
import android.widget.RelativeLayout
import io.plan8.backoffice.SharedPreferenceManager
import io.plan8.backoffice.adapter.RestfulAdapter
import io.plan8.backoffice.model.api.UploadInfo
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.http.Multipart
import java.io.File


/**
 * Created by chokwanghwan on 2017. 11. 9..
 */
class MoreFragment : BaseFragment() {
    var binding: FragmentMoreBinding? = null
    var progress: RelativeLayout? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val testData: MutableList<Any> = mutableListOf()
        //        이름, 전화번호, 주소, 예약일, 예약시간, 종료시간, 상품/이름, 추가 요청 사항, 내부 노트
        testData.add(MoreTitleItem("내 프로필"))
        var userName: String?
        if (Constants.me!!.userName != null) {
            userName = Constants.me!!.userName
        } else {
            userName = "이름 없음"
        }
        testData.add(MoreProfileItem(userName!!, Constants.me!!.phoneNumber!!))
        testData.add(MoreTitleItem("팀 선택"))

        if (Constants.me!!.team != null && Constants.me!!.team!!.isNotEmpty()) {
            for (team in Constants.me!!.team!!) {
                testData.add(MoreTeamItem(team.teamName, team.teamName))
            }
        }
        testData.add(EmptySpaceItem(0))
        binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.fragment_more, container, false)
        vm = MoreFragmentVM(this, savedInstanceState, testData)
        binding!!.setVariable(BR.vm, vm)
        binding!!.executePendingBindings()

        progress = binding!!.moreMenuProgressBar

        return binding!!.root
    }

    override fun onDestroy() {
        binding!!.unbind()
        super.onDestroy()
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        var vm: MoreFragmentVM? = null

        fun uploadImage(data: Uri, activity: Activity) {
            val uri = data
            val imagePath = getRealPathFromURI(uri, activity)

            val cursor = activity.contentResolver.query(Uri.parse(uri.toString()), null, null, null, null)
            cursor.moveToNext()
            val absolutePath = cursor.getString(cursor.getColumnIndex(MediaStore.MediaColumns.DATA))
            cursor.close()

            val files = File(absolutePath)
            val requestFile = RequestBody.create(MediaType.parse(activity.contentResolver.getType(uri)), files)
            val multipartBody: MultipartBody.Part = MultipartBody.Part.createFormData("files", files.name, requestFile)

            if (RestfulAdapter.instance!!.serviceApi != null) {
                RestfulAdapter.instance!!.serviceApi!!.postUpload("Bearer " + SharedPreferenceManager(activity).userToken, requestFile).enqueue(object : Callback<List<UploadInfo>> {
                    override fun onResponse(call: Call<List<UploadInfo>>?, response: Response<List<UploadInfo>>?) {
                        Log.e("test", "test")
                    }

                    override fun onFailure(call: Call<List<UploadInfo>>?, t: Throwable?) {
                        Log.e("test", "test")
                    }
                })
            }
        }

        fun getRealPathFromURI(imageUri: Uri, activity: Activity): String {
            val path = arrayOf(MediaStore.Images.Media.DATA)
            val cursor = activity.managedQuery(imageUri, path, null, null, null)
            val columnIndex = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA)
            cursor.moveToFirst()
            return cursor.getString(columnIndex)
        }
    }
}